### PR TITLE
feat: Allow payment cancels for more statuses

### PR DESCRIPTION
**Pull request by prajjwalkumar17**
requires_payment_method, requires_capture, requires_confirmation, requires_action

## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Previously while cancelling payment from some of the the payment intent statuses like requires_payment_method, requires_capture, requires_confirmation, requires_action there was an error thrown that these can't be cancelled as they haven't been authorized. Hence now we can cancel these payments too.
 
### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
As for the above statuses we aren't involving the connector and handling the requests ourselves in update method hence having a check and voding the payment attempt and cancelling the payment_intent status in DB.

## How did you test it?
Tested it by using postman as now we are able to cancel the payment in all those cases.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments.rs [hunk 1]
```diff
@@ -843,7 +843,9 @@ pub fn should_call_connector<Op: Debug, F: Clone>(
         }
         "PaymentCancel" => matches!(
             payment_data.payment_intent.status,
-            storage_enums::IntentStatus::RequiresCapture
+            storage_enums::IntentStatus::RequiresCapture // | storage_enums::IntentStatus::RequiresPaymentMethod
+                                                         // | storage_enums::IntentStatus::RequiresConfirmation
+                                                         // | storage_enums::IntentStatus::RequiresCustomerAction
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments.rs:848]: are these comments required?

File: crates/router/src/core/payments/operations/payment_cancel.rs [hunk 2]
```diff
@@ -170,18 +172,39 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsCancelRequest> for
...
+                    .update_payment_intent(
+                        payment_data.payment_intent,
+                        payment_intent_update,
+                        storage_scheme,
+                    )
+                    .await
+                    .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+                Some(enums::AttemptStatus::Voided)
+            } else {
+                Some(enums::AttemptStatus::VoidInitiated)
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_cancel.rs:190]: If this block always returns a value in both the branches, then this should not be optional

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/operations/payment_cancel.rs [hunk 3]
```diff
@@ -112,43 +112,45 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsCancelRequest>
...
+            &payment_intent.status,
+            &[
+                enums::IntentStatus::Failed,
+                enums::IntentStatus::Succeeded,
+                enums::IntentStatus::Cancelled,
+                enums::IntentStatus::Processing,
+                enums::IntentStatus::RequiresMerchantAction,
+            ],
+            "cancelled",
+        )?;
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_cancel.rs:125]: add this validation right after we get the `payment_intent` object. We do not need to query other entities if this function returns error.

File: crates/router/src/core/payments/operations/payment_cancel.rs [hunk 4]
```diff
@@ -170,18 +172,34 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsCancelRequest> for
...
+                    status: enums::IntentStatus::Cancelled,
+                };
+                payment_data.payment_intent = db
+                    .update_payment_intent(
+                        payment_data.payment_intent,
+                        payment_intent_update,
+                        storage_scheme,
+                    )
+                    .await
+                    .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_cancel.rs:187]: This block is doing two tasks. Make this block return the `intent_status` as an option and then do `async_map` on the status to update the `payment_intent`

File: crates/router/src/core/payments/operations/payment_cancel.rs [hunk 5]
```diff
@@ -172,18 +175,37 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsCancelRequest> for
...
-                payment_data.payment_attempt,
-                storage::PaymentAttemptUpdate::VoidUpdate {
-                    status: enums::AttemptStatus::VoidInitiated,
-                    cancellation_reason,
-                },
-                storage_scheme,
-            )
-            .await
-            .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+        let attempt_status_update =
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_cancel.rs:178]: Destructure this as a tuple of intent_status_update and attempt_status_update

**Review by Narayanbhat166 — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**